tls/s2n_post_handshake.c

```diff
@@ -20,59 +20,141 @@
 #include "tls/s2n_tls.h"
 #include "utils/s2n_safety.h"
 
-int s2n_post_handshake_recv(struct s2n_connection *conn) 
+/* The longest post-handshake message we currently support is NewSessionTicket.
+ * The only upper bound on session ticket sizes is UINT16_MAX.
+ * However, both s2n-tls and openssl tickets are generally on the order of a few hundred bytes.
+ * Therefore, let's use a reasonably small buffer and resize later if necessary.
+ */
+#define S2N_POST_HANDSHAKE_MESSAGE_LEN_ESTIMATE 2000
+
+static S2N_RESULT s2n_post_handshake_buffer_size(uint32_t message_len, uint32_t *buffer_size)
 {
-    POSIX_ENSURE_REF(conn);
+    if (message_len > 0) {
+        /* Allocate a little extra to account for the message header
+         * and to avoid reallocating if the next handshake message is slightly longer.
+         */
+        uint32_t extra = 100;
+
+        RESULT_GUARD_POSIX(s2n_add_overflow(message_len, extra, buffer_size));
+    } else {
+        *buffer_size = S2N_POST_HANDSHAKE_MESSAGE_LEN_ESTIMATE;
+    }
+    return S2N_RESULT_OK;
```

Comment:
Should there be a check here to ensure `message_len` is < `S2N_MAXIMUM_HANDSHAKE_MESSAGE_LENGTH`? It looks like `message_len` is taken from the message header in `s2n_try_read_full_handshake_message` -> `s2n_handshake_parse_header`, set to `buffer_size` here, and then passed to `s2n_stuffer_resize_if_empty` in `s2n_post_handshake_recv`. If there's no check for this, could an attacker not specify a huge message length that causes s2n-tls to run out of memory? Or am I missing a check for this somewhere?